### PR TITLE
Generalize CM::distribute_local_to_global for rectangular case.

### PR DESCRIPTION
**Pull request by kronbichler**
The new code allows to do the `distribute_local_to_global` operation when rows and columns are subject to different constraints, such as for flux matrices between different levels in multigrid methods. @guidokanschat this function should implement what we were looking for in case of the up/down flux matrices for multigrid for RT elements.

Note that this commit moves the `distribute_local_to_global` function to a different place in the code, but the only real change in the function is that we now use `col_constraint_matrix.make_sorted_row_list (col_indices, global_cols);` rather than `make_sorted_row_list (col_indices, global_cols);` (i.e., on the `*this` object).

### CODE REVIEW

**Comment by kronbichler** [comment]
This commit passes the test suite.

File: include/deal.II/lac/constraint_matrix.templates.h [hunk 1]
```diff
@@ -2541,6 +2487,76 @@ distribute_local_to_global (const FullMatrix<typename MatrixType::value_type>  &
...
+  const std::vector<size_type> &row_indices,
+  const ConstraintMatrix       &col_constraint_matrix,
+  const std::vector<size_type> &col_indices,
+  MatrixType                   &global_matrix) const
+{
+  typedef typename MatrixType::value_type number;
+
+  AssertDimension (local_matrix.m(), row_indices.size());
+  AssertDimension (local_matrix.n(), col_indices.size());
+  //Assert (sorted == true, ExcMatrixNotClosed());
```
Comment (tjhei) [inline include/deal.II/lac/constraint_matrix.templates.h]: can you remove this?

File: tests/lac/constraints_local_to_global_rect.cc [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+#include <deal.II/base/function.h>
+#include <deal.II/base/logstream.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/constraint_matrix.h>
+
+#include <fstream>
+#include <iostream>
+#include <complex>
+
+std::ofstream logfile("output");
```
Comment (tjhei) [inline tests/lac/constraints_local_to_global_rect.cc]: I don't think you need that line

**Comment by kronbichler** [comment]
@tjhei I fixed the code according to your remarks.

**Pull request merged**